vnet/conn_map.go

```diff
@@ -0,0 +1,135 @@
+package vnet
+
+import (
+	"fmt"
+	"net"
+)
+
+type udpConnMap struct {
+	portMap map[int][]*UDPConn
+}
+
+func newUDPConnMap() *udpConnMap {
+	return &udpConnMap{
+		portMap: map[int][]*UDPConn{},
+	}
+}
+
+func (m *udpConnMap) insert(conn *UDPConn) error {
+	udpAddr := conn.LocalAddr().(*net.UDPAddr)
+
+	// check if the port has a listener
+	conns, ok := m.portMap[udpAddr.Port]
+	if ok {
+		if udpAddr.IP.IsUnspecified() {
+			return fmt.Errorf("address already in use")
+		}
+
+		for _, conn := range conns {
+			laddr := conn.LocalAddr().(*net.UDPAddr)
+			if laddr.IP.IsUnspecified() {
```

Comment:
super nitpick feel free to ignore, I would prefer to put this and the if below together as `if laddr.IP.IsUnspecified() || laddr.IP.Equal(udpAddr.IP)` just because I was trying to understand why `IsUnspecified` will be a problem, then I realised (when reading the if below) that was a normal match

Reply:
Ok, that might be better!